runtime/ts/keymgmt/webcrypto.ts

```diff
@@ -0,0 +1,276 @@
+import {DeviceKey, Key, PrivateKey, PublicKey, RecoveryKey, SessionKey, WrappedKey} from "./keys";
+import {KeyGenerator, KeyStorage} from "./manager";
+import idb, {ObjectStore} from 'idb';
+import {decode, encode} from './base64';
+import rs from 'jsrsasign';
+
+class WebCryptoStorableKey<T> {
+
+    protected key: T;
+
+    constructor(key: T) {
+        this.key = key;
+    }
+
+    algorithm(): string {
+        // @ts-ignore
+        return (this.key as CryptoKey).algorithm.name;
+    }
+
+    storableKey(): T {
+        return this.key;
+    }
+}
+
+class WebCryptoWrappedKey implements WrappedKey {
+    private wrappedKeyData: Uint8Array;
+    private wrappedBy: PublicKey;
+
+    constructor(wrappedKeyData: Uint8Array, wrappedBy: PublicKey) {
+        this.wrappedKeyData = wrappedKeyData;
+        this.wrappedBy = wrappedBy;
+    }
+
+    algorithm(): string {
+        return this.wrappedBy.algorithm();
+    }
+
+    private unwrap(privKey: PrivateKey): PromiseLike<SessionKey> {
+        const webPrivKey = privKey as WebCryptoPrivateKey;
+
+        return crypto.subtle.unwrapKey(
+            "raw",
+            this.wrappedKeyData,
+            webPrivKey.cryptoKey(),
+            {
+                name: privKey.algorithm()
+            },
+            {
+                name: "AES-GCM",
+            },
+            true,
+            ["encrypt", "decrypt"]
+        ).then(key => new WebCryptoSessionKey(key));
+    }
+
+    rewrap(privKey: PrivateKey, pubKey: PublicKey): PromiseLike<WrappedKey> {
+        return this.unwrap(privKey).then(skey => skey.disposeToWrappedKeyUsing(pubKey));
+    }
+
+    export(): string {
+        return encode(this.wrappedKeyData.buffer as ArrayBuffer);
+    }
+
+    fingerprint(): PromiseLike<string> {
+        return Promise.resolve(encode(this.wrappedKeyData.buffer as ArrayBuffer));
+    }
+}
+
+class WebCryptoPrivateKey extends WebCryptoStorableKey<CryptoKey> implements PrivateKey {
+    constructor(key) {
+        super(key);
+    }
+
+    cryptoKey() {
+        return this.storableKey();
+    }
+
+    fingerprint(): PromiseLike<string> {
+        return Promise.resolve("");
+    }
+}
+
+class WebCryptoPublicKey extends WebCryptoStorableKey<CryptoKey> implements PublicKey {
+
+    constructor(key) {
+        super(key);
+    }
+
+    cryptoKey() {
+        return this.storableKey();
+    }
+
+    fingerprint(): PromiseLike<string> {
+        // TODO: fix this with a proper hash based fingerprint/thumbprint
+        return crypto.subtle.exportKey("jwk", this.cryptoKey()).then(key => JSON.stringify(key));
+    }
+}
+
+class WebCryptoSessionKey implements SessionKey {
```

Comment:
Consider restructuring somehow so that `encrypt` and `decrypt` are only exposed to unit test.